Switch to use enum class for typesetterOptions() template in ComplexTextControllerCoreText.mm

ddkilzer opened the pull request:
#### c2779697d1c92e035f4d5f0a0dc900c3021af691
<pre>
Switch to use enum class for typesetterOptions() template in ComplexTextControllerCoreText.mm
<a href="https://bugs.webkit.org/show_bug.cgi?id=273109">https://bugs.webkit.org/show_bug.cgi?id=273109</a>
&lt;<a href="https://rdar.apple.com/126898039">rdar://126898039</a>&gt;

Reviewed by Darin Adler.

* Source/WebCore/platform/graphics/mac/ComplexTextControllerCoreText.mm:
(WebCore::CoreTextTypesetterEmbeddingLevel): Add enum.
(WebCore::buildCoreTextTypesetterEmbeddingLevelDictionary): Add.
- Extract common code out of typesetterOptions().
(WebCore::typesetterOptions):
- Call buildCoreTextTypesetterEmbeddingLevelDictionary().
(WebCore::ComplexTextController::collectComplexTextRunsForCharacters):
- Switch from bool value to CoreTextTypesetterEmbeddingLevel enum.

Canonical link: <a href="https://commits.webkit.org/278294@main">https://commits.webkit.org/278294@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/03ffc6905cc78a6fc8de285845aa9dafbefca1eb

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/50117 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/29408 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/2410 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/53375 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/807 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/52420 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/35598 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/362 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/40902 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/52215 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/27072 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/43135 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/21993 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/24477 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/384 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/8494 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/46497 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/444 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/54956 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/25213 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/350 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/48306 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/26474 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/43318 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/47318 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/27337 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7241 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/26205 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/platform/graphics/mac/ComplexTextControllerCoreText.mm:
```diff
@@ -134,13 +134,15 @@
 }
 
-template<bool isLTR>
+enum class EmbeddingLevel : short { LTR = 0, RTL = 1 };
```
Change is fine as is, but seems like we would be better off making it clear that this is not just an arbitrary type, but must be `short` because of the use of `kCFNumberShortType` below, and have values that correspond to the `CoreText` embedding level.

We could deal with both by adding a brief comment, or do the two things by asserting the type is `short` right above the `optionValues` line of code with a `static_assert(std::is_same_v<short, decltype(embeddingLevelValue)>)` and naming the type `CoreTextEmbeddingLevel` or `CoreTextTypesetterEmbeddingLevel`.

Another variant would be to construct the value for the dictionary using the type safe Objective-C idiom: `@(enumToUnderlyingType(embeddingLevel))` and then using toll-free bridging. Then we would not need to use `short`, specifically. Could even build the whole dictionary that way with `@{ }`.

ddkilzer replied:
Fixed.

darinadler commented on Source/WebCore/platform/graphics/mac/ComplexTextControllerCoreText.mm:
```diff
@@ -134,13 +134,15 @@
+        short embeddingLevelValue = enumToUnderlyingType(embeddingLevel);
         const void* optionKeys[] = { kCTTypesetterOptionForcedEmbeddingLevel };
         const void* optionValues[] = { CFNumberCreate(kCFAllocatorDefault, kCFNumberShortType, &embeddingLevelValue) };
         options.construct(adoptCF(CFDictionaryCreate(kCFAllocatorDefault, optionKeys, optionValues, std::size(optionKeys), &kCFCopyStringDictionaryKeyCallBacks, &kCFTypeDictionaryValueCallBacks)));
```
No need to address this in this patch, but this seems unnecessarily inefficient. The actual creation of the value doesn’t need to be a template; the main value of the function template is the two separate globals. We could have the lambda call a non-template function passing arguments explicitly and save some code size.

ddkilzer replied:
I'll fix this as well.

Commit:
Switch to use enum class for typesetterOptions() template in ComplexTextControllerCoreText.mm
https://bugs.webkit.org/show_bug.cgi?id=273109
<rdar://126898039>

Reviewed by Darin Adler.

* Source/WebCore/platform/graphics/mac/ComplexTextControllerCoreText.mm:
(WebCore::CoreTextTypesetterEmbeddingLevel): Add enum.
(WebCore::buildCoreTextTypesetterEmbeddingLevelDictionary): Add.
- Extract common code out of typesetterOptions().
(WebCore::typesetterOptions):
- Call buildCoreTextTypesetterEmbeddingLevelDictionary().
(WebCore::ComplexTextController::collectComplexTextRunsForCharacters):
- Switch from bool value to CoreTextTypesetterEmbeddingLevel enum.

Canonical link: https://commits.webkit.org/278294@main
